packs/excel/actions/lib/excel_reader.py

```diff
@@ -0,0 +1,248 @@
+        keys = list(self._keys)
+        return keys
+
+    def get_variable_names(self):
+        ''' Returns the variable names in the sheet '''
+        if not self._ws:
+            raise KeyError("Sheet not specified")
+        variable_names = []
+        for col in range(self._variable_start_column,
+                         self._variable_end_column):
+
+            variable_names.append(self._ws.cell(column=col,
+                                  row=self._var_name_row).value)
+
+        return variable_names
+
+    def _set_key_column(self, key_column):
+        ''' Set the key column '''
+        self._key_column = key_column
+        self._keys = {}
+        row = self._data_start_row
+        while True:
+            key = self._ws.cell(column=self._key_column, row=row)
+            if key.value:
+                # Check for duplicate key
+                if key.value not in self._keys:
+                    self._keys[key.value] = row
+                else:
+                    self._unlock_file()
+                    raise ValueError("Duplicate key '%s' found at row '%s'." %
+                                     (key.value, row))
+            else:
+                self._data_end_row = row
+                break
+            row += 1
+
+    def _set_variable_start_column(self, variable_start_column):
+        ''' Set the template column in the template '''
+        col = self._variable_start_column = variable_start_column
+        while True:
+            cell = self._ws.cell(column=col, row=self._var_name_row)
+            if cell.value:
+                col += 1
+            else:
+                self._variable_end_column = col
+                break
+
+    def set_sheet(self, sheet_name, key_column=1, var_name_row=1,
+                  strict=False):
+        ''' Sets the current sheet '''
+        self._sheet_name = sheet_name
+        self._strict = strict
+        try:
+            self._ws = self._wb.get_sheet_by_name(self._sheet_name)
+        except KeyError:
+            if not self._strict:
+                if self._lock:
+                    self._ws = self._wb.create_sheet(self._sheet_name)
+                else:
+                    raise IOError("File not locked for modification")
+            else:
+                self._unlock_file()
+                raise KeyError("Sheet '%s' not found" % self._sheet_name)
+
+        self._var_name_row = var_name_row
+        self._data_start_row = var_name_row + 1
+
+        self._key_column = key_column
+        self._set_key_column(key_column)
+
+        self._variable_start_column = key_column + 1
+        self._variable_end_column = self._variable_start_column
+        self._set_variable_start_column(key_column + 1)
+
+    def get_row_for_key(self, key):
+        ''' Returns the row for a given key, or -1 if not matched '''
+        if len(key) > 255:
+            self._unlock_file()
+            raise ValueError("Key exceeds 255 characters")
+        if key in self._keys:
```

Comment:
I 'll probably go for a `return self._keys.get(key)` which will return None if key is not found and then in set_values_for_variables, test if row is not None instead of testing if row equals -1